tnefparse/cmdline.py

```diff
@@ -57,8 +57,7 @@ def tnefparse():
     args = parser.parse_args()
 
     if args.logging:
-        level = eval("logging." + args.logging)
-        logger.setLevel(level)
+        logger.setLevel(getattr(logging, args.logging))
 
```

Comment:
These lines are currently not under test.

Would you mind to add a quick test?

You could have a look at https://github.com/koodaamo/tnefparse/pull/83/files for a similar test setup.

Reply:
If I write a new test:

`ret = script_runner.run('tnefparse', '--logging', 'INFO', '-rb', 'tests/examples/rtf.tnef')`

it will have empty `ret.stderr`, although I'd expect the two lines of log there. Where are they gone?

Reply:
Can you please link to your branch?

Reply:
See https://github.com/eumiro/tnefparse/blob/eval-is-evil/tests/test_cmdline.py#L87

Reply:
So :-) This one line change had some effects :-)

I am not super adept with configuring logging, so maybe there is something going on.

In order to limit the problem scope I tried some things...

- run you command with the real app, and redirect stderr into a file: worked! it is there!
- in order to exclude `pytest-console-scripts` as a problem, I tried `pytest` only (testing `tnefparse` function directly) - also, nothing in `cap = capsys.readouterr() -> cap.err`
- I then ran pytest completely (no breakpoint), and wth... I got the following output
```
-------------------------------------------- Captured stderr call --------------------------------------------
2020-11-27 14:43:42,140 - tnefparse - ERROR - before
2020-11-27 14:43:42,140 - tnefparse - ERROR - after
2020-11-27 14:43:42,140 - tnefparse - INFO - Skipping checksum for performance
========================================== short test summary info ===========================================
FAILED tests/test_cmdline.py::test_cmdline_logging_info[inprocess] - assert False
====================================== 1 failed, 29 deselected in 0.14s ======================================
ERROR: InvocationError for command /home/jugmac00/Projects/tnef-parallel/tnefparse/.tox/py38/bin/pytest -k test_cmdline_logging_info (exited with code 1)
__________________________________________________ summary ___________________________________________________
ERROR:   py38: commands failed
```

As I said, I am neither very deep into Python logging, nor into `pytest` internals, but this pretty much looks like a pytest bug.

Maybe it is this one https://github.com/pytest-dev/pytest/issues/5997 ?

@asottile By chance, you got a minute to have a look at this?

Reply:
`logging.basicConfig(...)` is called globally, this attaches it to some output object early on in the process.  pytest tries its best to replace the output from the `sys` module but if things have references to the old stdout / stderr from import side-effects then they'll use those (you can confirm this by moving the logging statements inside the function)

this might be how I would write that test to test the logging btw:

```diff
diff --git a/tests/test_cmdline.py b/tests/test_cmdline.py
index 99b682f..5995a39 100644
--- a/tests/test_cmdline.py
+++ b/tests/test_cmdline.py
@@ -1,4 +1,5 @@
 import json
+import logging
 import shutil
 import sys
 import tempfile
@@ -84,8 +85,11 @@ def test_handle_value_errors(mocked_TNEF):
         tnefparse()
 
 
-def test_cmdline_logging_info(script_runner):
-    ret = script_runner.run('tnefparse', '-l', 'INFO', '-rb', 'tests/examples/rtf.tnef')
-    assert len(ret.stdout) == 593
-    assert any(line.startswith("INFO") for line in ret.stderr.splitlines())
-    assert ret.success
+def test_cmdline_logging_info(caplog, capsys):
+    retv = tnefparse.cmdline.tnefparse(('-l', 'INFO', '-rb', 'tests/examples/rtf.tnef'))
+    assert not retv
+    stdout, _ = capsys.readouterr()
+    assert len(stdout) == 593
+    assert caplog.record_tuples == [
+        ('tnefparse', logging.INFO, 'Skipping checksum for performance'),
+    ]
diff --git a/tnefparse/cmdline.py b/tnefparse/cmdline.py
index cf49e8f..ce977f8 100644
--- a/tnefparse/cmdline.py
+++ b/tnefparse/cmdline.py
@@ -47,14 +47,14 @@ argument('-d', '--dump', action="store_true", default=False,
          help="extract a json dump of the tnef contents")
 
 
-def tnefparse():
+def tnefparse(argv=None):
     "command-line script"
 
     if len(sys.argv) == 1:
         parser.print_help()
         sys.exit(1)
 
-    args = parser.parse_args()
+    args = parser.parse_args(argv)
 
     if args.logging:
         logger.setLevel(getattr(logging, args.logging))
```

(I'm unsubscribing from this thread, if you need me, @ - mention me directly)

Reply:
Thank you, Anthony, now it works. I was able to add further tests and also simplify the `argparse/logging` in `tnefparse/cmdline.py`.

Reply:
@eumiro I guess #81 was closed unintentionally. I assumed it was merged already.